File: src/lib/stores/spoolman.ts
import { invoke } from '@tauri-apps/api/core';
import { writable } from 'svelte/store';

export interface SpoolmanFilament {
  id: string;
  manufacturer: string;
  name: string;
  material: string;
  density: number;
  diameter: number;
  weight: number | null;
  spool_weight: number | null;
  color_hex: string | null;
  extruder_temp: number | null;
  bed_temp: number | null;
  extruder_temp_range: number[] | null;
  bed_temp_range: number[] | null;
  translucent: boolean;
  glow: boolean;
}

export interface SpoolmanResponse {
  items: SpoolmanFilament[];
  total: number;
}

export interface ColorGroup {
  hex: string;
  filaments: SpoolmanFilament[];
  materials: Set<string>;
  brands: Set<string>;
  hasTranslucent: boolean;
  hasGlow: boolean;
}

let cachedBrands: string[] | null = null;
let cachedMaterials: string[] | null = null;
let cachedColors: Map<string, ColorGroup> | null = null;

export const spoolmanFilaments = writable<SpoolmanFilament[]>([]);
export const spoolmanTotal = writable<number>(0);
export const spoolmanLoading = writable<boolean>(false);
export const spoolmanBrands = writable<string[]>([]);
export const spoolmanMaterials = writable<string[]>([]);
export const spoolmanColors = writable<Map<string, ColorGroup>>(new Map());

export async function searchSpoolman(
  query?: string,
  vendor?: string,
  material?: string,
  limit: number = 50,
  offset: number = 0
) {
  spoolmanLoading.set(true);
  try {
    const result = await invoke<SpoolmanResponse>('search_spoolman', {
      query: query || null,
      vendor: vendor || null,
      material: material || null,
      limit,
      offset,
    });
    
    if (offset === 0) {
      spoolmanFilaments.set(result.items);
    } else {
      spoolmanFilaments.update(items => [...items, ...result.items]);
    }
    
    spoolmanTotal.set(result.total);
  } catch (error) {
    console.error('Failed to search SpoolmanDB:', error);
    throw error;
  } finally {
    spoolmanLoading.set(false);
  }
}

export async function loadBrands() {
  if (cachedBrands) {
    spoolmanBrands.set(cachedBrands);
    return;
  }
  
  try {
    const brands = await invoke<string[]>('get_spoolman_brands');
    cachedBrands = brands;
    spoolmanBrands.set(brands);
  } catch (error) {
    console.error('Failed to load brands:', error);
  }
}

export async function loadMaterials() {
  if (cachedMaterials) {
    spoolmanMaterials.set(cachedMaterials);
    return;
  }
  
  try {
    const materials = await invoke<string[]>('get_spoolman_materials');
    cachedMaterials = materials;
    spoolmanMaterials.set(materials);
  } catch (error) {
    console.error('Failed to load materials:', error);
  }
}

export async function loadColors() {
  if (cachedColors) {
    console.log('Using cached colors');
    spoolmanColors.set(cachedColors);
    return;
  }
  
  try {
    const result = await invoke<SpoolmanResponse>('search_spoolman', {
      query: null,
      vendor: null,
      material: null,
      limit: 10000,
      offset: 0,
    });
    
    const colorMap = new Map<string, ColorGroup>();
    
    result.items.forEach(filament => {
      if (!filament.color_hex) return;
      
      const hex = filament.color_hex.replace('#', '').toLowerCase();
      
      if (!colorMap.has(hex)) {
        colorMap.set(hex, {
          hex: hex,
          filaments: [],
          materials: new Set(),
          brands: new Set(),
          hasTranslucent: false,
          hasGlow: false,
        });
      }
      
      const group = colorMap.get(hex)!;
      group.filaments.push(filament);
      group.materials.add(filament.material);
      group.brands.add(filament.manufacturer);
      if (filament.translucent) group.hasTranslucent = true;
      if (filament.glow) group.hasGlow = true;
    });
    
    console.log(`✅ Loaded ${colorMap.size} unique colors from SpoolmanDB`);
    cachedColors = colorMap;
    spoolmanColors.set(colorMap);
  } catch (error) {
    console.error('Failed to load colors:', error);
  }
}

export async function syncSpoolmanDB() {
  try {
    await invoke('sync_spoolman_db');
    cachedBrands = null;
    cachedMaterials = null;
    cachedColors = null;
    await loadBrands();
    await loadMaterials();
    await loadColors();
  } catch (error) {
    console.error('Failed to sync SpoolmanDB:', error);
    throw error;
  }
}

export async function debugFilament(filament: SpoolmanFilament) {
  try {
    await invoke('debug_filament', { filament });
  } catch (error) {
    console.error('Debug failed:', error);
  }
}

File: src-tauri/src/spoolman.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::sync::Mutex;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SpoolmanFilament {
    pub id: String,
    pub manufacturer: String,
    pub name: String,
    pub material: String,
    pub density: f64,
    pub diameter: f64,
    #[serde(default)]
    pub color_hex: Option<String>,
    #[serde(default)]
    pub color_hexes: Option<Vec<String>>,
    #[serde(default)]
    pub weight: Option<f64>,
    #[serde(default)]
    pub spool_weight: Option<f64>,
    #[serde(default)]
    pub spool_type: Option<String>,
    #[serde(rename = "extruder_temp", default)]
    pub extruder_temp: Option<i32>,
    #[serde(rename = "extruder_temp_range", default)]
    pub extruder_temp_range: Option<Vec<i32>>,
    #[serde(rename = "bed_temp", default)]
    pub bed_temp: Option<i32>,
    #[serde(rename = "bed_temp_range", default)]
    pub bed_temp_range: Option<Vec<i32>>,
    #[serde(default)]
    pub finish: Option<String>,
    #[serde(default)]
    pub multi_color_direction: Option<String>,
    #[serde(default)]
    pub pattern: Option<String>,
    #[serde(default)]
    pub translucent: bool,
    #[serde(default)]
    pub glow: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SpoolmanResponse {
    pub items: Vec<SpoolmanFilament>,
    pub total: usize,
}

pub struct SpoolmanClient {
    client: reqwest::Client,
    base_url: String,
    cache: Mutex<Option<Vec<SpoolmanFilament>>>,
}

impl SpoolmanClient {
    pub fn new() -> Self {
        Self {
            client: reqwest::Client::new(),
            base_url: "https://donkie.github.io/SpoolmanDB".to_string(),
            cache: Mutex::new(None),
        }
    }

    async fn ensure_cache(&self) -> Result<(), String> {
        let mut cache = self.cache.lock().unwrap();
        
        if cache.is_some() {
            println!("✅ Using cached SpoolmanDB data");
            return Ok(());
        }

        drop(cache);

        let url = format!("{}/filaments.json", self.base_url);
        println!("📥 Downloading SpoolmanDB from: {}", url);

        let response = self
            .client
            .get(&url)
            .send()
            .await
            .map_err(|e| format!("Failed to fetch: {}", e))?;

        println!("✅ SpoolmanDB response status: {}", response.status());

        let filaments: Vec<SpoolmanFilament> = response
            .json()
            .await
            .map_err(|e| {
                eprintln!("JSON parse error: {:?}", e);
                format!("Failed to parse JSON: {}", e)
            })?;

        println!("✅ Cached {} filaments from SpoolmanDB", filaments.len());

        let mut cache = self.cache.lock().unwrap();
        *cache = Some(filaments);

        Ok(())
    }

    pub async fn search_filaments(
        &self,
        query: Option<String>,
        vendor: Option<String>,
        material: Option<String>,
        limit: usize,
        offset: usize,
    ) -> Result<SpoolmanResponse, String> {
        self.ensure_cache().await?;

        let cache = self.cache.lock().unwrap();
        let mut filaments = cache.as_ref().unwrap().clone();
        drop(cache);

        if let Some(q) = query {
            let q_lower = q.to_lowercase();
            filaments.retain(|f| {
                f.name.to_lowercase().contains(&q_lower)
                    || f.manufacturer.to_lowercase().contains(&q_lower)
                    || f.material.to_lowercase().contains(&q_lower)
            });
        }

        if let Some(v) = vendor {
            let v_lower = v.to_lowercase();
            filaments.retain(|f| f.manufacturer.to_lowercase() == v_lower);
        }

        if let Some(m) = material {
            let m_lower = m.to_lowercase();
            filaments.retain(|f| f.material.to_lowercase() == m_lower);
        }

        let total = filaments.len();
        let items = filaments
            .into_iter()
            .skip(offset)
            .take(limit)
            .collect::<Vec<_>>();

        Ok(SpoolmanResponse { items, total })
    }

    pub async fn get_brands(&self) -> Result<Vec<String>, String> {
        self.ensure_cache().await?;

        let cache = self.cache.lock().unwrap();
        let filaments = cache.as_ref().unwrap();

        let mut brands: Vec<String> = filaments
            .iter()
            .map(|f| f.manufacturer.clone())
            .collect::<HashSet<_>>()
            .into_iter()
            .collect();

        brands.sort();
        println!("✅ Loaded {} unique brands from cache", brands.len());

        Ok(brands)
    }

    pub async fn get_materials(&self) -> Result<Vec<String>, String> {
        self.ensure_cache().await?;

        let cache = self.cache.lock().unwrap();
        let filaments = cache.as_ref().unwrap();

        let materials_set: HashSet<String> = filaments
            .iter()
            .map(|f| f.material.clone())
            .collect();

        let mut materials: Vec<String> = materials_set.into_iter().collect();
        materials.sort();

        println!("✅ Loaded {} unique materials from cache", materials.len());

        Ok(materials)
    }

    pub async fn sync_database(&self) -> Result<(), String> {
        println!("🔄 Force syncing SpoolmanDB...");
        let mut cache = self.cache.lock().unwrap();
        *cache = None;
        drop(cache);
        
        self.ensure_cache().await?;
        println!("✅ SpoolmanDB synced successfully");
        Ok(())
    }
}
